staging/src/k8s.io/client-go/tools/cache/controller_test.go

```diff
@@ -893,3 +906,219 @@ func (m *mockTxnStore) Transaction(txns ...Transaction) *TransactionError {
 	m.succeedCount = len(txns)
 	return nil
 }
+
+func TestAtomicReplaceEquivalence(t *testing.T) {
+	_, ctx := ktesting.NewTestContext(t)
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Run a scenario with pod creates/deletes/updates with a relist with simple replace.
+	simpleScenario := setupMockReplaceTester(t, false)
+	testReplaceScenario(t, ctx, simpleScenario)
+	// Obtain the events and items from the store after the scenario.
+	eventsDisabled := simpleScenario.getHistory()
+	itemsDisabled := simpleScenario.getStoredItems()
+
+	simpleScenario.teardown()
+
+	// Run a scenario with pod creates/deletes/updates with a relist with atomic replace.
+	atomicScenario := setupMockReplaceTester(t, true)
+	testReplaceScenario(t, ctx, atomicScenario)
+	// Obtain the events and items from the store after the scenario.
+	eventsEnabled := atomicScenario.getHistory()
+	itemsEnabled := atomicScenario.getStoredItems()
+
+	atomicScenario.teardown()
+
+	// Compare Results
+	// The final state of the store must be identical
+	if !apiequality.Semantic.DeepEqual(itemsEnabled, itemsDisabled) {
```

Comment:
To make the test more explicit and easier to follow it might be better to validate against hardcoded results, instead of comparing. 

Simplified version
```
func TestControllerOperations(t *testing.T) {
  for _, atomic := []bool{true, false} {
    t.Run(fmt.Sprintf("Atomic=%v", atomic), func (t *testing.T) {
      controller := NewController(atomic)
      testAtomicReplaceEquivalence(controller)
    })
  }
} 
func testControllerOperations(t *testing.T, c *controller) {
...
}
```

Reply:
This was a great point, we don't need to check against each other and it acts as a test in case the underlying implementation breaks both versions.